// @antv/l7-layers/lib/core/interface.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.lineStyleType = exports.TextureBlend = exports.SizeUnitType = exports.LinearDir = exports.CanvasUpdateType = void 0;
var lineStyleType;
exports.lineStyleType = lineStyleType;

(function (lineStyleType) {
  lineStyleType[lineStyleType["solid"] = 0] = "solid";
  lineStyleType[lineStyleType["dash"] = 1] = "dash";
})(lineStyleType || (exports.lineStyleType = lineStyleType = {}));

var LinearDir;
exports.LinearDir = LinearDir;

(function (LinearDir) {
  LinearDir["VERTICAL"] = "vertical";
  LinearDir["HORIZONTAL"] = "horizontal";
})(LinearDir || (exports.LinearDir = LinearDir = {}));

var TextureBlend;
/**
 * 基础图层类型定义
 */

exports.TextureBlend = TextureBlend;

(function (TextureBlend) {
  TextureBlend["NORMAL"] = "normal";
  TextureBlend["REPLACE"] = "replace";
})(TextureBlend || (exports.TextureBlend = TextureBlend = {}));

var SizeUnitType;
exports.SizeUnitType = SizeUnitType;

(function (SizeUnitType) {
  SizeUnitType[SizeUnitType["pixel"] = 0] = "pixel";
  SizeUnitType[SizeUnitType["meter"] = 1] = "meter";
})(SizeUnitType || (exports.SizeUnitType = SizeUnitType = {}));

var CanvasUpdateType;
exports.CanvasUpdateType = CanvasUpdateType;

(function (CanvasUpdateType) {
  CanvasUpdateType["ALWAYS"] = "always";
  CanvasUpdateType["DRAGEND"] = "dragend";
})(CanvasUpdateType || (exports.CanvasUpdateType = CanvasUpdateType = {}));
