// @antv/l7-renderer/lib/regl/ReglModel.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/regl/ReglModel.ts
var ReglModel_exports = {};
__export(ReglModel_exports, {
  default: () => ReglModel
});
module.exports = __toCommonJS(ReglModel_exports);
var import_l7_core = require("@antv/l7-core");
var import_lodash = require("lodash");
var import_constants = require("./constants");
var ReglModel = class {
  constructor(reGl, options) {
    this.destroyed = false;
    this.uniforms = {};
    this.reGl = reGl;
    const {
      vs,
      fs,
      attributes,
      uniforms,
      primitive,
      count,
      elements,
      depth,
      cull,
      instances
    } = options;
    const reglUniforms = {};
    this.options = options;
    if (uniforms) {
      this.uniforms = this.extractUniforms(uniforms);
      Object.keys(uniforms).forEach((uniformName) => {
        reglUniforms[uniformName] = reGl.prop(uniformName);
      });
    }
    const reglAttributes = {};
    Object.keys(attributes).forEach((name) => {
      reglAttributes[name] = attributes[name].get();
    });
    const drawParams = {
      attributes: reglAttributes,
      frag: fs,
      uniforms: reglUniforms,
      vert: vs,
      colorMask: reGl.prop("colorMask"),
      blend: {
        enable: reGl.prop("blend.enable"),
        func: reGl.prop("blend.func"),
        equation: reGl.prop("blend.equation"),
        color: reGl.prop("blend.color")
      },
      stencil: {
        enable: reGl.prop("stencil.enable"),
        mask: reGl.prop("stencil.mask"),
        func: reGl.prop("stencil.func"),
        opFront: reGl.prop("stencil.opFront"),
        opBack: reGl.prop("stencil.opBack")
      },
      primitive: import_constants.primitiveMap[primitive === void 0 ? import_l7_core.gl.TRIANGLES : primitive]
    };
    if (instances) {
      drawParams.instances = instances;
    }
    if (count) {
      drawParams.count = count;
    } else if (elements) {
      drawParams.elements = elements.get();
    }
    this.initDepthDrawParams({ depth }, drawParams);
    this.initCullDrawParams({ cull }, drawParams);
    this.drawCommand = reGl(drawParams);
    this.drawParams = drawParams;
  }
  updateAttributesAndElements(attributes, elements) {
    const reglAttributes = {};
    Object.keys(attributes).forEach((name) => {
      reglAttributes[name] = attributes[name].get();
    });
    this.drawParams.attributes = reglAttributes;
    this.drawParams.elements = elements.get();
    this.drawCommand = this.reGl(this.drawParams);
  }
  updateAttributes(attributes) {
    const reglAttributes = {};
    Object.keys(attributes).forEach((name) => {
      reglAttributes[name] = attributes[name].get();
    });
    this.drawParams.attributes = reglAttributes;
    this.drawCommand = this.reGl(this.drawParams);
  }
  addUniforms(uniforms) {
    this.uniforms = {
      ...this.uniforms,
      ...this.extractUniforms(uniforms)
    };
  }
  draw(options, pick) {
    if (this.drawParams.attributes && Object.keys(this.drawParams.attributes).length === 0) {
      return;
    }
    const uniforms = {
      ...this.uniforms,
      ...this.extractUniforms(options.uniforms || {})
    };
    const reglDrawProps = {};
    Object.keys(uniforms).forEach((uniformName) => {
      const type = typeof uniforms[uniformName];
      if (type === "boolean" || type === "number" || Array.isArray(uniforms[uniformName]) || uniforms[uniformName].BYTES_PER_ELEMENT) {
        reglDrawProps[uniformName] = uniforms[uniformName];
      } else {
        reglDrawProps[uniformName] = uniforms[uniformName].get();
      }
    });
    reglDrawProps.blend = pick ? this.getBlendDrawParams({
      blend: { enable: false }
    }) : this.getBlendDrawParams(options);
    reglDrawProps.stencil = this.getStencilDrawParams(options);
    reglDrawProps.colorMask = this.getColorMaskDrawParams(options, pick);
    this.drawCommand(reglDrawProps);
  }
  destroy() {
    var _a, _b;
    (_b = (_a = this.drawParams) == null ? void 0 : _a.elements) == null ? void 0 : _b.destroy();
    if (this.options.attributes) {
      Object.values(this.options.attributes).forEach((attr) => {
        attr == null ? void 0 : attr.destroy();
      });
    }
    this.destroyed = true;
  }
  initDepthDrawParams({ depth }, drawParams) {
    if (depth) {
      drawParams.depth = {
        enable: depth.enable === void 0 ? true : !!depth.enable,
        mask: depth.mask === void 0 ? true : !!depth.mask,
        func: import_constants.depthFuncMap[depth.func || import_l7_core.gl.LESS],
        range: depth.range || [0, 1]
      };
    }
  }
  getBlendDrawParams({
    blend
  }) {
    const { enable, func, equation, color = [0, 0, 0, 0] } = blend || {};
    return {
      enable: !!enable,
      func: {
        srcRGB: import_constants.blendFuncMap[func && func.srcRGB || import_l7_core.gl.SRC_ALPHA],
        srcAlpha: import_constants.blendFuncMap[func && func.srcAlpha || import_l7_core.gl.SRC_ALPHA],
        dstRGB: import_constants.blendFuncMap[func && func.dstRGB || import_l7_core.gl.ONE_MINUS_SRC_ALPHA],
        dstAlpha: import_constants.blendFuncMap[func && func.dstAlpha || import_l7_core.gl.ONE_MINUS_SRC_ALPHA]
      },
      equation: {
        rgb: import_constants.blendEquationMap[equation && equation.rgb || import_l7_core.gl.FUNC_ADD],
        alpha: import_constants.blendEquationMap[equation && equation.alpha || import_l7_core.gl.FUNC_ADD]
      },
      color
    };
  }
  getStencilDrawParams({
    stencil
  }) {
    const {
      enable,
      mask = -1,
      func = {
        cmp: import_l7_core.gl.ALWAYS,
        ref: 0,
        mask: -1
      },
      opFront = {
        fail: import_l7_core.gl.KEEP,
        zfail: import_l7_core.gl.KEEP,
        zpass: import_l7_core.gl.KEEP
      },
      opBack = {
        fail: import_l7_core.gl.KEEP,
        zfail: import_l7_core.gl.KEEP,
        zpass: import_l7_core.gl.KEEP
      }
    } = stencil || {};
    return {
      enable: !!enable,
      mask,
      func: {
        ...func,
        cmp: import_constants.stencilFuncMap[func.cmp]
      },
      opFront: {
        fail: import_constants.stencilOpMap[opFront.fail],
        zfail: import_constants.stencilOpMap[opFront.zfail],
        zpass: import_constants.stencilOpMap[opFront.zpass]
      },
      opBack: {
        fail: import_constants.stencilOpMap[opBack.fail],
        zfail: import_constants.stencilOpMap[opBack.zfail],
        zpass: import_constants.stencilOpMap[opBack.zpass]
      }
    };
  }
  getColorMaskDrawParams({ stencil }, pick) {
    const colorMask = (stencil == null ? void 0 : stencil.enable) && stencil.opFront && !pick ? [false, false, false, false] : [true, true, true, true];
    return colorMask;
  }
  initCullDrawParams({ cull }, drawParams) {
    if (cull) {
      const { enable, face = import_l7_core.gl.BACK } = cull;
      drawParams.cull = {
        enable: !!enable,
        face: import_constants.cullFaceMap[face]
      };
    }
  }
  extractUniforms(uniforms) {
    const extractedUniforms = {};
    Object.keys(uniforms).forEach((uniformName) => {
      this.extractUniformsRecursively(uniformName, uniforms[uniformName], extractedUniforms, "");
    });
    return extractedUniforms;
  }
  extractUniformsRecursively(uniformName, uniformValue, uniforms, prefix) {
    if (uniformValue === null || typeof uniformValue === "number" || typeof uniformValue === "boolean" || Array.isArray(uniformValue) && typeof uniformValue[0] === "number" || (0, import_lodash.isTypedArray)(uniformValue) || uniformValue === "" || "resize" in uniformValue) {
      uniforms[`${prefix && prefix + "."}${uniformName}`] = uniformValue;
      return;
    }
    if ((0, import_lodash.isPlainObject)(uniformValue)) {
      Object.keys(uniformValue).forEach((childName) => {
        this.extractUniformsRecursively(childName, uniformValue[childName], uniforms, `${prefix && prefix + "."}${uniformName}`);
      });
    }
    if (Array.isArray(uniformValue)) {
      uniformValue.forEach((child, idx) => {
        Object.keys(child).forEach((childName) => {
          this.extractUniformsRecursively(childName, child[childName], uniforms, `${prefix && prefix + "."}${uniformName}[${idx}]`);
        });
      });
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
